### PR TITLE
Allow the user to specify a tenant when acquiring a token to talk to …

### PR DESCRIPTION
**Pull request by mingweiw**
…the BF connectors

### CODE REVIEW

File: libraries/Microsoft.Bot.Connector/Authentication/MicrosoftAppCredentials.cs [hunk 1]
```diff
@@ -65,10 +65,20 @@ public MicrosoftAppCredentials(string appId, string password)
         /// <param name="appId">The Microsoft app ID.</param>
         /// <param name="password">The Microsoft app password.</param>
         /// <param name="customHttpClient">Optional <see cref="HttpClient"/> to be used when acquiring tokens.</param>
-        public MicrosoftAppCredentials(string appId, string password, HttpClient customHttpClient)
+        public MicrosoftAppCredentials(string appId, string password, HttpClient customHttpClient) : this(appId, password, null, null) { }
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Connector/Authentication/MicrosoftAppCredentials.cs]: BUG. This is incorrect. If I read the code right, this should be 
: this (appId, password, null, customHttpClient) {}

File: libraries/Microsoft.Bot.Connector/Authentication/MicrosoftAppCredentials.cs [hunk 2]
```diff
@@ -95,15 +105,31 @@ public MicrosoftAppCredentials(string appId, string password, HttpClient customH
         /// </value>
         public string MicrosoftAppPassword { get; set; }
 
+        /// <summary>
+        /// Gets or sets the channel auth token tenant for this credential.
+        /// </summary>
+        private string _channelAuthTenant;
+        public string ChannelAuthTenant
+        {
+            get { return string.IsNullOrEmpty(_channelAuthTenant) ? "botframework.com" : _channelAuthTenant; }
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Connector/Authentication/MicrosoftAppCredentials.cs]: Please move "botframework.com" to the constants file.

File: libraries/Microsoft.Bot.Connector/Authentication/MicrosoftAppCredentials.cs [hunk 3]
```diff
@@ -95,15 +105,31 @@ public MicrosoftAppCredentials(string appId, string password, HttpClient customH
...
-        /// </value>
         public virtual string OAuthEndpoint
         {
-            get { return AuthenticationConstants.ToChannelFromBotLoginUrl; }
+            get
+            {
+                // Advanced user only, see https://aka.ms/bots/tenant-restriction
+                var endpointUrl = string.Format(AuthenticationConstants.ToChannelFromBotLoginUrlTemplate, ChannelAuthTenant);
+
+                if (Uri.TryCreate(endpointUrl, UriKind.Absolute, out var result))
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Connector/Authentication/MicrosoftAppCredentials.cs]: This doesn't pass stylecop. Please make sure this builds with no warnings.

File: libraries/Microsoft.Bot.Connector/Authentication/MicrosoftAppCredentials.cs [hunk 4]
```diff
@@ -95,15 +105,31 @@ public MicrosoftAppCredentials(string appId, string password, HttpClient customH
...
+                    return endpointUrl;
+
+                throw new Exception($"Invalid token endpoint: {endpointUrl}");
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Connector/Authentication/MicrosoftAppCredentials.cs]: This is not a useful Exception. I would *NEVER* expect to get this on a "Get" property. This should be thown when the property is set.

File: tests/Microsoft.Bot.Connector.Tests/GetTokenRefreshTests.cs [hunk 5]
```diff
@@ -25,6 +25,14 @@ public async Task TokenTests_GetCredentialsWorks()
             Assert.NotNull(result);
         }
 
+        [Fact]
+        public async Task TokenTests_GetCredentialsFromTenantWorks()
+        {
+            MicrosoftAppCredentials credentials = new MicrosoftAppCredentials("645cd89f-a83e-4af9-abb5-a454e917cbc4", "jvoMWRBA67:zjgePZ359_-_", "microsoft.com", null);
```
Comment (cleemullins) [inline tests/Microsoft.Bot.Connector.Tests/GetTokenRefreshTests.cs]: 2 things here:
1. We cannot put magic keys / constants in code. If we did this, the keys needs to be stored as encrypted variables in Azure Devops and passed into the build via ENV variables. (This is a super-common pattern). 

2. We cannot put actual "Service" tests here. These are unit tests, and cannot hit actual services.

File: libraries/Microsoft.Bot.Connector/Authentication/MicrosoftAppCredentials.cs (hunk 1 above)
Comment (mingweiw) [reply libraries/Microsoft.Bot.Connector/Authentication/MicrosoftAppCredentials.cs]: doh!

File: libraries/Microsoft.Bot.Connector/Authentication/MicrosoftAppCredentials.cs (hunk 2 above)
Comment (mingweiw) [reply libraries/Microsoft.Bot.Connector/Authentication/MicrosoftAppCredentials.cs]: yep

File: libraries/Microsoft.Bot.Connector/Authentication/MicrosoftAppCredentials.cs (hunk 3 above)
Comment (mingweiw) [reply libraries/Microsoft.Bot.Connector/Authentication/MicrosoftAppCredentials.cs]: k :)

File: libraries/Microsoft.Bot.Connector/Authentication/MicrosoftAppCredentials.cs (hunk 4 above)
Comment (mingweiw) [reply libraries/Microsoft.Bot.Connector/Authentication/MicrosoftAppCredentials.cs]: good point.

File: tests/Microsoft.Bot.Connector.Tests/GetTokenRefreshTests.cs (hunk 5 above)
Comment (mingweiw) [reply tests/Microsoft.Bot.Connector.Tests/GetTokenRefreshTests.cs]: this is how tests are written in this file. I'm copying the existing pattern.

Comment (cleemullins) [reply tests/Microsoft.Bot.Connector.Tests/GetTokenRefreshTests.cs]: Alright, I'll clean this up seperatly.

**Pull request merged**